cmd/clusterctl/pkg/client/cluster/components.go

```diff
@@ -49,43 +63,106 @@ func (p *providerComponents) Create(components repository.Components) error {
 	resources := sortResourcesForCreate(components.Objs())
 
 	// creates (or updates) provider components
-	for _, r := range resources {
+	for i := range resources {
+		obj := resources[i]
 
 		// check if the component already exists, and eventually update it
 		currentR := &unstructured.Unstructured{}
-		currentR.SetGroupVersionKind(r.GroupVersionKind())
+		currentR.SetGroupVersionKind(obj.GroupVersionKind())
 
 		key := client.ObjectKey{
-			Namespace: r.GetNamespace(),
-			Name:      r.GetName(),
+			Namespace: obj.GetNamespace(),
+			Name:      obj.GetName(),
 		}
 		if err = c.Get(ctx, key, currentR); err != nil {
 			if !apierrors.IsNotFound(err) {
 				return errors.Wrapf(err, "failed to get current provider object")
 			}
 
 			//if it does not exists, create the component
-			klog.V(3).Infof("Creating: %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
-			if err = c.Create(ctx, &r); err != nil { //nolint
-				return errors.Wrapf(err, "failed to create provider object %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+			klog.V(3).Infof("Creating: %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
+			if err = c.Create(ctx, &obj); err != nil {
+				return errors.Wrapf(err, "failed to create provider object %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
 			}
 
 			continue
 		}
 
 		// otherwise update the component
-		klog.V(3).Infof("Updating: %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+		klog.V(3).Infof("Updating: %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
 
 		// if upgrading an existing component, then use the current resourceVersion for the optimistic lock
-		r.SetResourceVersion(currentR.GetResourceVersion())
-		if err = c.Update(ctx, &r); err != nil { //nolint
-			return errors.Wrapf(err, "failed to update provider object %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+		obj.SetResourceVersion(currentR.GetResourceVersion())
+		if err = c.Update(ctx, &obj); err != nil {
+			return errors.Wrapf(err, "failed to update provider object %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
 		}
 	}
 
 	return nil
 }
 
+func (p *providerComponents) Delete(options DeleteOptions) error {
+	// Fetch all the components belonging to a provider.
+	labels := map[string]string{
+		clusterctlv1.ClusterctlProviderLabelName: options.Provider.Name,
```

Comment:
Sidenote, we should probably change the name of this const now that the label is generic

Reply:
Yep. I have to align clusterctl to the changes on labels, but this touches many places so I will fix it in another PR